File: scripts/lib/team_plan.py
#!/usr/bin/env python3
"""team_plan.py — expand a /team plan.json into per-subtask files + a manifest.

Usage:
    team_plan.py <plan.json> <workdir>

plan.json is a JSON array of subtasks:
  [{"label": "build-ui", "task": "<text>", "backend": "agy"|"codex"|"native", "tier": "standard"}]

For each subtask this writes the raw task text to <workdir>/<idx>.task (so the task is
NEVER passed through a shell — team.sh feeds the file to run.sh on stdin, injection-safe),
and prints one manifest line per subtask to stdout:

  <AGY|CODEX|NATIVE>\t<idx>\t<label>\t<tier>\t<workdir>/<idx>.task

backend is normalized: gemini/agy -> AGY; codex/chatgpt/openai -> CODEX; claude/native/sonnet/opus
-> NATIVE. Backends are equal — any CLI backend dispatches via run.sh; NATIVE is solved by Claude.
Entries missing a task are skipped (noted on stderr). Exit 0 unless the plan is unreadable.
"""
import json
import os
import re
import sys

AGY = {"agy", "gemini", "flash", "pro", "google"}
CODEX = {"codex", "chatgpt", "openai", "gpt"}
NATIVE = {"native", "claude", "sonnet", "opus", "anthropic"}
CLI_TIERS = {"cheap", "standard"}      # any CLI backend (agy, codex, ...)
NATIVE_TIERS = {"sonnet", "opus"}


def _tier(raw, backend):
    # Allowlist the tier per backend. This is also the security boundary: only these
    # fixed values reach the TAB/NEWLINE-delimited manifest, so a crafted tier with an
    # embedded \t/\n can never forge an extra manifest row (TSV injection).
    raw = str(raw or "").strip().lower()
    if backend == "NATIVE":
        return raw if raw in NATIVE_TIERS else "sonnet"
    return raw if raw in CLI_TIERS else "standard"


def _sanitize_label(label, idx):
    label = re.sub(r"[^A-Za-z0-9._-]+", "-", str(label or "").strip()).strip("-")
    return label[:48] or f"task{idx}"


def _backend(name):
    name = str(name or "").strip().lower()
    if name in AGY:
        return "AGY"
    if name in CODEX:
        return "CODEX"
    if name in NATIVE:
        return "NATIVE"
    return ""


def main():
    if len(sys.argv) < 3:
        sys.stderr.write("usage: team_plan.py <plan.json> <workdir>\n")
        return 2
    plan_path, workdir = sys.argv[1], sys.argv[2]
    try:
        with open(plan_path, encoding="utf-8") as f:
            plan = json.load(f)
    except Exception as e:  # noqa: BLE001
        sys.stderr.write(f"team_plan: cannot read plan: {e}\n")
        return 2
    if not isinstance(plan, list):
        sys.stderr.write("team_plan: plan must be a JSON array\n")
        return 2

    os.makedirs(workdir, exist_ok=True)
    out = []
    for idx, item in enumerate(plan):
        if not isinstance(item, dict):
            sys.stderr.write(f"team_plan: skip non-object entry #{idx}\n")
            continue
        task = item.get("task")
        if not task or not str(task).strip():
            sys.stderr.write(f"team_plan: skip entry #{idx} with empty task\n")
            continue
        backend = _backend(item.get("backend", "native"))
        if not backend:
            backend = "NATIVE"  # unknown -> safe default
        tier = _tier(item.get("tier"), backend)
        label = _sanitize_label(item.get("label", ""), idx)
        taskfile = os.path.join(workdir, f"{idx}.task")
        with open(taskfile, "w", encoding="utf-8", newline="\n") as tf:
            tf.write(str(task))
        out.append(f"{backend}\t{idx}\t{label}\t{tier}\t{taskfile}")

    sys.stdout.write("\n".join(out) + ("\n" if out else ""))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/lib/team-plan.mjs
/**
 * team-plan.mjs — expand a /team plan.json into per-subtask files + a manifest.
 * Node ESM port of scripts/lib/team_plan.py. Zero runtime dependencies.
 *
 * Exports:
 *   planToManifest(plan, workdir)
 *     plan    — array of { label, task, backend, tier, deps?, verify? }
 *     workdir — directory to write <idx>.task files into (created if absent)
 *     returns — string[] of TSV rows: "BE\tidx\tlabel\ttier\tpath"
 *
 * Backend normalisation: agy/gemini/flash/pro/google -> AGY
 *                        codex/chatgpt/openai/gpt    -> CODEX
 *                        claude/native/sonnet/opus/anthropic -> NATIVE
 *                        unknown                     -> NATIVE  (safe default)
 * Tier allowlist + TSV-injection hardening: only whitelisted values reach the manifest;
 * embedded \t / \n in a crafted tier string are neutralised before any comparison.
 * Label sanitised to [A-Za-z0-9._-] ≤48 chars.
 * Empty-task entries are skipped (noted to stderr).
 */

import { mkdirSync, writeFileSync } from 'node:fs';
import { join }                     from 'node:path';

// ─── constants ───────────────────────────────────────────────────────────────

const AGY    = new Set(['agy', 'gemini', 'flash', 'pro', 'google']);
const CODEX  = new Set(['codex', 'chatgpt', 'openai', 'gpt']);
const NATIVE = new Set(['native', 'claude', 'sonnet', 'opus', 'anthropic']);

const CLI_TIERS    = new Set(['cheap', 'standard']);
const NATIVE_TIERS = new Set(['sonnet', 'opus']);

// ─── helpers ─────────────────────────────────────────────────────────────────

function _normalizeBackend(name) {
  const lc = String(name ?? '').trim().toLowerCase();
  if (AGY.has(lc))    return 'AGY';
  if (CODEX.has(lc))  return 'CODEX';
  if (NATIVE.has(lc)) return 'NATIVE';
  return '';  // unknown
}

/**
 * Allowlist the tier per backend.  Also the TSV-injection boundary:
 * strip \t and \n BEFORE comparison so a crafted tier can't forge a new row.
 */
function _normalizeTier(raw, backend) {
  const t = String(raw ?? '').replace(/[\t\n\r]/g, '').trim().toLowerCase();
  if (backend === 'NATIVE') return NATIVE_TIERS.has(t) ? t : 'sonnet';
  return CLI_TIERS.has(t) ? t : 'standard';
}

function _sanitizeLabel(label, idx) {
  const clean = String(label ?? '').trim().replace(/[^A-Za-z0-9._-]+/g, '-').replace(/^-+|-+$/g, '');
  return (clean.slice(0, 48)) || `task${idx}`;
}

// ─── planToManifest ──────────────────────────────────────────────────────────

/**
 * Write per-subtask .task files and return TSV manifest rows.
 *
 * @param {object[]} plan     Array of subtask objects from plan.json
 * @param {string}   workdir  Directory for .task files (created if absent)
 * @returns {string[]}        TSV rows: "BE\tidx\tlabel\ttier\tpath"
 */
export function planToManifest(plan, workdir) {
  if (!Array.isArray(plan)) {
    throw new TypeError('planToManifest: plan must be an array');
  }

  mkdirSync(workdir, { recursive: true });

  const rows = [];

  for (let idx = 0; idx < plan.length; idx++) {
    const item = plan[idx];

    if (!item || typeof item !== 'object') {
      process.stderr.write(`team_plan: skip non-object entry #${idx}\n`);
      continue;
    }

    const taskText = item.task != null ? String(item.task) : '';
    if (!taskText.trim()) {
      process.stderr.write(`team_plan: skip entry #${idx} with empty task\n`);
      continue;
    }

    let backend = _normalizeBackend(item.backend ?? 'native');
    if (!backend) backend = 'NATIVE';  // unknown -> safe default

    const tier  = _normalizeTier(item.tier, backend);
    const label = _sanitizeLabel(item.label, idx);

    // Write raw task text, LF line endings, no shell processing.
    const taskPath = join(workdir, `${idx}.task`);
    writeFileSync(taskPath, taskText.replace(/\r\n/g, '\n').replace(/\r/g, '\n'), { encoding: 'utf8' });

    rows.push(`${backend}\t${idx}\t${label}\t${tier}\t${taskPath}`);
  }

  return rows;
}
